src/Framework/Framework/Controls/ItemsControl.cs

```diff
@@ -90,32 +90,36 @@ public static void OnCompilation(ResolvedControl control, BindingCompilationServ
-                    new BindingParameterAnnotation(dataContext, new CurrentCollectionIndexExtensionParameter())))));
+                parsedExpression: CreateIndexBindingExpression(dataContext))));
         }
 
-        protected IBinding GetIndexBinding(IDotvvmRequestContext context)
+        private static ParameterExpression CreateIndexBindingExpression(DataContextStack dataContext) =>
+            Expression.Parameter(typeof(int), "_index")
+                .AddParameterAnnotation(new BindingParameterAnnotation(dataContext, new CurrentCollectionIndexExtensionParameter()));
+
+        // TODO: change this to int.ToString(CultureInfo.InvariantCulture) in a separate PR
+        private static readonly Expression<Func<int, string>> IndexToStringNoGlobalizeExpression = (int i) => IndexToStringNoGlobalize(i);
+        internal static MethodInfo IndexToStringNoGlobalizeMethod => ((MethodCallExpression)IndexToStringNoGlobalizeExpression.Body).Method;
+        internal static string IndexToStringNoGlobalize(int i) => i.ToString();
+
+        protected ValueOrBinding<string> GetIndexClientIdFragment(IDotvvmRequestContext context)
         {
-            var result = GetValueRaw(Internal.CurrentIndexBindingProperty) as IBinding;
-            if (result is {})
-            {
-                return result;
-            }
-            else
+            var binding = GetValueRaw(Internal.CurrentIndexBindingProperty) as IBinding;
+            if (binding == null)
             {
-                // slower path: create the _index binding at runtime
+                // slower path: create the _index binding at runtime and cache it
                 var bindingService = context.Services.GetRequiredService<BindingCompilationService>();
                 var dataContext = GetDataSourceBinding().GetProperty<CollectionElementDataContextBindingProperty>().DataContext;
-                return bindingService.Cache.CreateCachedBinding("_index", new object[] { dataContext }, () =>
-                    new ValueBindingExpression<int>(bindingService, new object?[] {
+                binding = bindingService.Cache.CreateCachedBinding(
+                    "_index", new object[] { dataContext },
+                    () => new ValueBindingExpression<string>(bindingService, new object?[] {
                         dataContext,
-                        new ParsedExpressionBindingProperty(
-                            Expression.Parameter(typeof(int), "_index").AddParameterAnnotation(
-                                new BindingParameterAnnotation(dataContext, new CurrentCollectionIndexExtensionParameter()))
-                        )
-                    }
-                ));
+                        new ParsedExpressionBindingProperty(CreateIndexBindingExpression(dataContext))
+                    }));
             }
+
+            return new ValueOrBinding<int>(binding).Select(i => IndexToStringNoGlobalize(i));
```

Comment:
I'm really not a fan of using the Select in core components at runtime. I don't know why can't we leave this logic as it was before, insisting on using `ValueOrBinding<string>` internally causes more problems than good things...